### PR TITLE
Remi database

### PR DESCRIPTION
**Pull request by Flechman**
Finished the interface to interact with the database, and tested it. The interface provides a basic set of methods that should be (for the moment) enough to store and retrieve storables (only Event for the moment).
The interface is equiped with other requirements for the storables, such as a storer for each storable class. This is necessary to work in a relational maner with the database.

### CODE REVIEW

**Review by Sosol02 — DISMISSED** [review]
lgtm

**Comment by Sosol02** [comment]
hmm en fait non j'ai rien dit

**Review by hugolan — APPROVED** [review]

File: app/src/androidTest/java/com/github/onedirection/testdatabase/ConcreteDatabaseTest.java [hunk 1]
```diff
@@ -2,41 +2,184 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
 @RunWith(AndroidJUnit4.class)
 public class ConcreteDatabaseTest {
 
-    private ArrayList<Item> makeItems(int count) {
-        ArrayList<Item> l = new ArrayList<>();
+    private int count = 50;
+    private int emptyCoordProportion = 2; //Proportion of empty coordinates events will be of 1/emptyCoordProportion
```
Comment (Ef55) [inline app/src/androidTest/java/com/github/onedirection/testdatabase/ConcreteDatabaseTest.java:30]: Should be static :)

File: app/src/main/java/com/github/onedirection/EventQueries.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    public List<Event> getEventsByWeek() {
+        //TODO
+        return null;
+    }
+
+    public List<Event> getEventsByMonth() {
+        //TODO
+        return null;
+    }
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/EventQueries.java:29]: Missing implementations (but apparently it is another task).

File: app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java [hunk 3]
```diff
@@ -46,24 +53,21 @@ private ConcreteDatabase() {
         return result;
     }
 
+    @Override
     public <T extends Storable<T>> CompletableFuture<T> retrieve(Id id, Storer<T> storer) {
+        if(storer == null) { throw new IllegalArgumentException("storer is null"); }
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java:58]: Should use 

```suggestion
     Objects.requireNonNull(storer, "Storer is null");
```

File: app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java [hunk 4]
```diff
@@ -74,4 +78,103 @@ private ConcreteDatabase() {
         return result;
     }
 
+    @Override
+    public <T extends Storable<T>> CompletableFuture<Id> remove(Id id, Storer<T> storer) {
+        if(storer == null) { throw new IllegalArgumentException("storer is null"); }
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/ConcreteDatabase.java:83]: Same here; use _requireNonNull_

File: app/src/main/java/com/github/onedirection/database/store/EventStorer.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+public class EventStorer extends Storer<Event> {
+
+    private static final EventStorer GLOBAL = new EventStorer();
+    public static final String KEY_ID = "id";
+    public static final String KEY_NAME = "name";
+    public static final String KEY_COORD_LATITUDE = "coordLatitude";
+    public static final String KEY_COORD_LONGITUDE = "coordLongitude";
+    public static final String KEY_COORD_NAME = "locationName";
+    public static final String KEY_EPOCH_START_TIME = "epochStartTime";
+    public static final String KEY_EPOCH_END_TIME = "epochEndTime";
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/store/EventStorer.java:28]: Using an _enum_ and an [EnumMap](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/EnumMap.html) would be (imo) better code, but it is already good enough as-is.

File: app/src/main/java/com/github/onedirection/database/store/EventStorer.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public Class<Event> classTag() {
+        return Event.class;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    @Override
+    public Map<String, Object> storableToMap(Event storable) {
+        if(storable == null) {
```
Comment (Ef55) [inline app/src/main/java/com/github/onedirection/database/store/EventStorer.java:47]: Again, _requireNonNull_

**Review by Ef55 — APPROVED** [review]
Overall a really nice implementation.

File: app/src/main/java/com/github/onedirection/database/store/EventStorer.java (hunk 6 above)
Comment (Ef55) [reply app/src/main/java/com/github/onedirection/database/store/EventStorer.java:47]: (And the RequiresApi should go away, didn't realized).

**Pull request merged**